### PR TITLE
Add duck-duck-go-search extension

### PR DESCRIPTION
**Pull request by tegola**
## Description

Adds support for Duck Duck Go search with suggestions, with the option to save the search history locally.

Sorry for the messed up commit history, the `ray` command was continuosly asking me to pull contributions I had already merged, I had to recreate my own `extensions` repository fork to fix it.

## Screencast

duck-duck-go-search 2023-04-10 at 14 36 53

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/duck-duck-go-search/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                      onAction={async () => {
+                        await deleteHistoryItem(item);
+                      }}
+                      icon={{ source: Icon.Trash }}
+                      shortcut={{ modifiers: ["cmd"], key: "d" }}
+                    />
+                  )}
+
+                  <Action
+                    title="Clear All History"
```
Comment (pernielsentikaer) [inline extensions/duck-duck-go-search/src/index.tsx:45]:
```suggestion
                    title="Clear All History"
                    style={Action.Style.Destructive}
```

File: extensions/duck-duck-go-search/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                  />
+
+                  <Action.CopyToClipboard title="Copy URL to Clipboard" content={item.url} />
+                  <Action.CopyToClipboard title="Copy Suggestion to Clipboard" content={item.query} />
+                </ActionPanel.Section>
+
+                <ActionPanel.Section title="History">
+                  {item.isHistory && (
+                    <Action
+                      title="Remove From History"
```
Comment (pernielsentikaer) [inline extensions/duck-duck-go-search/src/index.tsx:35]:
```suggestion
                      title="Remove from History"
                      style={Action.Style.Destructive}
```

File: extensions/duck-duck-go-search/src/utils/useSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    setHistory(newHistory);
+
+    if (rememberSearchHistory) {
+      await LocalStorage.setItem(HISTORY_KEY, JSON.stringify(newHistory));
+    }
+  }
+
+  async function deleteAllHistory() {
+    await LocalStorage.removeItem(HISTORY_KEY);
```
Comment (pernielsentikaer) [inline extensions/duck-duck-go-search/src/utils/useSearch.tsx:102]: I would implement [confirmAlert](https://developers.raycast.com/api-reference/feedback/alert#confirmalert) here

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi 👋 

Thanks for your contribution 💪 

I have now tested your extension, and I have some feedback ready for you:

* Seems like some throttle could be a good idea, it's bouncing a bit - not sure if it's possible to improve.

I'm looking forward to testing this extension again 🔥 

Request a new review when you are ready. Feel free to contact me here or at [Slack](https://www.raycast.com/community) if you have any questions.

**Comment by tegola** [comment]
Thanks for your suggestions, which I have implemented. I also changed the shortcuts and style/icons for deleting history items to better match with other plugins using an history (clipboard, calculator).

I didn't implement throttling because search seemed a little unresponsive.

**Review by pernielsentikaer — APPROVED** [review]

**Comment by justspacedog** [comment]
I had the same approch but with a little bit more functionality build in here: [https://github.com/raycast/extensions/pull/5754](https://github.com/raycast/extensions/pull/5754)

Maybe you can add this? I closed the pull request because I want to make sure there is no copyright on the icon, but it seems that everybody here doesn't care (and legal@duckduckgo.com doesn't answer me).

**Comment by tegola** [comment]
I'm not an avid bangs user, and I feel they break the one-functionality-per-plugin approach in Raycast. Also, DuckDuckGo suggestions aren't even really related to the selected bang. But! If you think they're useful and like to contribute to my release, please do! And don't forget to add your name to the credits!

Regarding the copyright on the icon, I made a little research and ultimately just decided to use it thinking I'd be covered by a universal acceptable use policy.

**Comment by justspacedog** [comment]
> I'm not an avid bangs user, and I feel they break the one-functionality-per-plugin approach in Raycast. Also, DuckDuckGo suggestions aren't even really related to the selected bang. But! If you think they're useful and like to contribute to my release, please do! And don't forget to add your name to the credits!

I use bangs all the time and think it's a good idea to actually see where you're searching. But you are right that duckduckgo's suggestions don't always match the suggestions on the other website.

> Regarding the copyright on the icon, I made a little research and ultimately just decided to use it thinking I'd be covered by a universal acceptable use policy.

I don't know. But I'd better ask, because I've heard a lot of strange things.

I hope I have done it right, here is the update: [https://github.com/raycast/extensions/pull/6062](https://github.com/raycast/extensions/pull/6062)

**Comment by tegola** [comment]
> I hope I have done it right, here is the update: #6062

Your pull request actually replaced all files, but as long the unmodified files’ content is the same, I guess it’s ok?

I’m away from my computer, but I noticed it’s already been merged, so please make sure the published extension works as intended.

Thanks

**Comment by tegola** [comment]
My bad, it hasn’t been merged. Cramped screen fooled me.

I’ll take a better look tomorrow, thanks.

**Pull request merged**